File: blocker.js
// blocker.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');
const { execSync } = require('child_process');

const COLORS = {
    green: '\x1b[92m',
    red: '\x1b[91m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    reset: '\x1b[0m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const CONFIG_DIR = path.join(os.homedir(), '.blocker');
const CONFIG_FILE = path.join(CONFIG_DIR, 'config.json');
const HOSTS_FILE = os.platform() === 'win32'
    ? 'C:\\Windows\\System32\\drivers\\etc\\hosts'
    : '/etc/hosts';
const BLOCK_IP = '127.0.0.1';

function ensureConfig() {
    if (!fs.existsSync(CONFIG_DIR)) fs.mkdirSync(CONFIG_DIR, { recursive: true });
    if (!fs.existsSync(CONFIG_FILE)) {
        const defaultConfig = { blocked: [], active: false, whitelist: [] };
        fs.writeFileSync(CONFIG_FILE, JSON.stringify(defaultConfig, null, 2));
    }
}

function loadConfig() {
    ensureConfig();
    return JSON.parse(fs.readFileSync(CONFIG_FILE, 'utf8'));
}

function saveConfig(cfg) {
    fs.writeFileSync(CONFIG_FILE, JSON.stringify(cfg, null, 2));
}

function readHosts() {
    return fs.readFileSync(HOSTS_FILE, 'utf8').split('\n');
}

function writeHosts(lines) {
    fs.writeFileSync(HOSTS_FILE, lines.join('\n'));
}

function updateHosts(domains, action) {
    let lines = readHosts();
    const filtered = lines.filter(line => {
        for (const d of domains) {
            if (line.trim().startsWith(BLOCK_IP) && line.includes(d)) return false;
        }
        return true;
    });
    if (action === 'add') {
        for (const d of domains) {
            filtered.push(`${BLOCK_IP} ${d}`);
        }
    }
    writeHosts(filtered);
}

function startBlocking(domains, durationMinutes) {
    const cfg = loadConfig();
    cfg.blocked = [...new Set([...cfg.blocked, ...domains])];
    if (durationMinutes) {
        cfg.timerEnd = new Date(Date.now() + durationMinutes * 60000).toISOString();
    }
    cfg.active = true;
    saveConfig(cfg);
    try {
        updateHosts(domains, 'add');
        console.log(colorize(`Blocked ${domains.length} domains.`, 'green'));
        if (durationMinutes) {
            console.log(colorize(`Blocking active for ${durationMinutes} min.`, 'yellow'));
            setTimeout(() => {
                console.log(colorize('Time expired. Unblocking...', 'yellow'));
                disableBlocking();
            }, durationMinutes * 60000);
        }
    } catch (err) {
        console.log(colorize(`Error: ${err.message}`, 'red'));
        process.exit(1);
    }
}

function disableBlocking() {
    const cfg = loadConfig();
    const blocked = cfg.blocked;
    if (blocked.length === 0) {
        console.log(colorize('No active blocking.', 'yellow'));
        return;
    }
    try {
        updateHosts(blocked, 'remove');
        cfg.blocked = [];
        cfg.active = false;
        cfg.timerEnd = null;
        saveConfig(cfg);
        console.log(colorize('Blocking disabled.', 'green'));
    } catch (err) {
        console.log(colorize(`Error: ${err.message}`, 'red'));
    }
}

function showStatus() {
    const cfg = loadConfig();
    if (cfg.active && cfg.timerEnd) {
        const remaining = Math.floor((new Date(cfg.timerEnd) - Date.now()) / 1000);
        if (remaining > 0) {
            const mins = Math.floor(remaining / 60);
            const secs = remaining % 60;
            console.log(colorize(`Blocking active. Remaining: ${mins} min ${secs} sec`, 'blue'));
        } else {
            console.log(colorize('Blocking active but timer expired.', 'yellow'));
        }
    } else if (cfg.active) {
        console.log(colorize('Blocking active (no timer).', 'blue'));
    } else {
        console.log(colorize('Blocking inactive.', 'yellow'));
    }
    if (cfg.blocked.length > 0) {
        console.log(colorize(`Blocked domains (${cfg.blocked.length}):`, 'green'));
        cfg.blocked.forEach(d => console.log(`  - ${d}`));
    } else {
        console.log(colorize('No blocked domains.', 'yellow'));
    }
}

function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.log(colorize('Usage: blocker add|remove|list|enable|disable|status|pomodoro|clear [domains...] [options]', 'yellow'));
        process.exit(1);
    }
    const cmd = args[0];
    const rest = args.slice(1);

    ensureConfig();
    const cfg = loadConfig();

    switch (cmd) {
        case 'add': {
            if (rest.length === 0) {
                console.log(colorize('Specify domains to block.', 'yellow'));
                return;
            }
            const timeIdx = rest.indexOf('-t');
            let duration = 0;
            if (timeIdx !== -1 && rest.length > timeIdx+1) {
                duration = parseInt(rest[timeIdx+1]);
                rest.splice(timeIdx, 2);
            }
            startBlocking(rest, duration);
            break;
        }
        case 'remove': {
            if (rest.length === 0) {
                console.log(colorize('Specify domains to remove.', 'yellow'));
                return;
            }
            const blocked = cfg.blocked;
            for (const d of rest) {
                const idx = blocked.indexOf(d);
                if (idx !== -1) blocked.splice(idx, 1);
            }
            saveConfig(cfg);
            try {
                updateHosts(rest, 'remove');
                console.log(colorize(`Removed: ${rest.join(', ')}`, 'green'));
            } catch (err) {
                console.log(colorize(`Error: ${err.message}`, 'red'));
            }
            break;
        }
        case 'list': {
            const blocked = cfg.blocked;
            if (blocked.length === 0) {
                console.log(colorize('No blocked domains.', 'yellow'));
            } else {
                console.log(colorize('Blocked domains:', 'green'));
                blocked.forEach(d => console.log(`  - ${d}`));
            }
            break;
        }
        case 'enable': {
            if (cfg.blocked.length === 0) {
                console.log(colorize('No domains in blocklist. Add them with "add".', 'yellow'));
                return;
            }
            const timeIdx = rest.indexOf('-t');
            let duration = 0;
            if (timeIdx !== -1 && rest.length > timeIdx+1) {
                duration = parseInt(rest[timeIdx+1]);
            }
            startBlocking(cfg.blocked, duration);
            break;
        }
        case 'disable': {
            disableBlocking();
            break;
        }
        case 'status': {
            showStatus();
            break;
        }
        case 'pomodoro': {
            if (cfg.blocked.length === 0) {
                console.log(colorize('No domains in blocklist. Add them with "add".', 'yellow'));
                return;
            }
            console.log(colorize('Starting Pomodoro: 25 minutes work', 'blue'));
            startBlocking(cfg.blocked, 25);
            setTimeout(() => {
                console.log(colorize('Pomodoro finished! 5 minutes break.', 'green'));
                disableBlocking();
                setTimeout(() => {
                    console.log(colorize('Break over. Start next Pomodoro.', 'blue'));
                }, 5 * 60000);
            }, 25 * 60000);
            break;
        }
        case 'clear': {
            const force = rest.includes('-f');
            if (!force) {
                const readline = require('readline').createInterface({
                    input: process.stdin,
                    output: process.stdout
                });
                readline.question(colorize('Clear all blocklist? [y/N] ', 'yellow'), ans => {
                    readline.close();
                    if (ans.toLowerCase() === 'y') doClear();
                });
            } else {
                doClear();
            }
            break;
        }
        default:
            console.log(colorize(`Unknown command: ${cmd}`, 'red'));
    }
}

function doClear() {
    const cfg = loadConfig();
    if (cfg.blocked.length === 0) {
        console.log(colorize('Blocklist already empty.', 'yellow'));
        return;
    }
    try {
        updateHosts(cfg.blocked, 'remove');
        cfg.blocked = [];
        cfg.active = false;
        cfg.timerEnd = null;
        saveConfig(cfg);
        console.log(colorize('Blocklist cleared.', 'green'));
    } catch (err) {
        console.log(colorize(`Error: ${err.message}`, 'red'));
    }
}

main();

File: blocker.rb
#!/usr/bin/env ruby
# blocker.rb
# encoding: UTF-8

require 'json'
require 'fileutils'
require 'time'

COLORS = {
  green: "\e[92m",
  red: "\e[91m",
  yellow: "\e[93m",
  blue: "\e[94m",
  reset: "\e[0m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

CONFIG_DIR = File.join(Dir.home, '.blocker')
CONFIG_FILE = File.join(CONFIG_DIR, 'config.json')
HOSTS_FILE = RUBY_PLATFORM =~ /mswin|mingw|cygwin/ ?
  'C:\Windows\System32\drivers\etc\hosts' : '/etc/hosts'
BLOCK_IP = '127.0.0.1'

def ensure_config
  FileUtils.mkdir_p(CONFIG_DIR)
  unless File.exist?(CONFIG_FILE)
    default = { 'blocked' => [], 'active' => false, 'whitelist' => [] }
    File.write(CONFIG_FILE, JSON.pretty_generate(default))
  end
end

def load_config
  ensure_config
  JSON.parse(File.read(CONFIG_FILE))
end

def save_config(cfg)
  File.write(CONFIG_FILE, JSON.pretty_generate(cfg))
end

def read_hosts
  File.readlines(HOSTS_FILE).map(&:chomp)
rescue Errno::EACCES
  raise "Требуются права администратора для изменения hosts-файла"
end

def write_hosts(lines)
  File.write(HOSTS_FILE, lines.join("\n") + "\n")
rescue Errno::EACCES
  raise "Требуются права администратора для изменения hosts-файла"
end

def update_hosts(domains, action)
  lines = read_hosts
  new_lines = lines.reject do |line|
    domains.any? { |d| line.strip.start_with?(BLOCK_IP) && line.include?(d) }
  end
  if action == 'add'
    domains.each { |d| new_lines << "#{BLOCK_IP} #{d}" }
  end
  write_hosts(new_lines)
end

def start_blocking(domains, minutes)
  cfg = load_config
  cfg['blocked'] = (cfg['blocked'] + domains).uniq
  if minutes && minutes > 0
    cfg['timer_end'] = (Time.now + minutes * 60).iso8601
  end
  cfg['active'] = true
  save_config(cfg)
  update_hosts(domains, 'add')
  puts colorize("Blocked #{domains.size} domains.", :green)
  if minutes && minutes > 0
    puts colorize("Blocking active for #{minutes} min.", :yellow)
    Thread.new { timer_thread(minutes) }
  end
end

def timer_thread(minutes)
  sleep(minutes * 60)
  puts colorize("Time expired. Unblocking...", :yellow)
  disable_blocking
end

def disable_blocking
  cfg = load_config
  if cfg['blocked'].empty?
    puts colorize("No active blocking.", :yellow)
    return
  end
  update_hosts(cfg['blocked'], 'remove')
  cfg['blocked'] = []
  cfg['active'] = false
  cfg['timer_end'] = nil
  save_config(cfg)
  puts colorize("Blocking disabled.", :green)
end

def show_status
  cfg = load_config
  if cfg['active'] && cfg['timer_end']
    remaining = (Time.iso8601(cfg['timer_end']) - Time.now).to_i
    if remaining > 0
      mins = remaining / 60
      secs = remaining % 60
      puts colorize("Blocking active. Remaining: #{mins} min #{secs} sec", :blue)
    else
      puts colorize("Blocking active but timer expired.", :yellow)
    end
  elsif cfg['active']
    puts colorize("Blocking active (no timer).", :blue)
  else
    puts colorize("Blocking inactive.", :yellow)
  end
  if cfg['blocked'].any?
    puts colorize("Blocked domains (#{cfg['blocked'].size}):", :green)
    cfg['blocked'].each { |d| puts "  - #{d}" }
  else
    puts colorize("No blocked domains.", :yellow)
  end
end

def main
  if ARGV.empty?
    puts colorize("Usage: blocker add|remove|list|enable|disable|status|pomodoro|clear [domains...] [options]", :yellow)
    exit 1
  end

  cmd = ARGV[0]
  rest = ARGV[1..-1] || []
  cfg = load_config

  case cmd
  when 'add'
    if rest.empty?
      puts colorize("Specify domains to block.", :yellow)
      return
    end
    minutes = 0
    idx = rest.index('-t')
    if idx && rest[idx+1]
      minutes = rest[idx+1].to_i
      rest.delete_at(idx)
      rest.delete_at(idx)
    end
    start_blocking(rest, minutes)

  when 'remove'
    if rest.empty?
      puts colorize("Specify domains to remove.", :yellow)
      return
    end
    cfg['blocked'] -= rest
    save_config(cfg)
    update_hosts(rest, 'remove')
    puts colorize("Removed: #{rest.join(', ')}", :green)

  when 'list'
    if cfg['blocked'].empty?
      puts colorize("No blocked domains.", :yellow)
    else
      puts colorize("Blocked domains:", :green)
      cfg['blocked'].each { |d| puts "  - #{d}" }
    end

  when 'enable'
    if cfg['blocked'].empty?
      puts colorize("No domains in blocklist. Add them with 'add'.", :yellow)
      return
    end
    minutes = 0
    idx = rest.index('-t')
    if idx && rest[idx+1]
      minutes = rest[idx+1].to_i
    end
    start_blocking(cfg['blocked'].dup, minutes)

  when 'disable'
    disable_blocking

  when 'status'
    show_status

  when 'pomodoro'
    if cfg['blocked'].empty?
      puts colorize("No domains in blocklist. Add them with 'add'.", :yellow)
      return
    end
    puts colorize("Starting Pomodoro: 25 minutes work", :blue)
    start_blocking(cfg['blocked'].dup, 25)
    sleep(25 * 60)
    puts colorize("Pomodoro finished! 5 minutes break.", :green)
    disable_blocking
    sleep(5 * 60)
    puts colorize("Break over. Start next Pomodoro.", :blue)

  when 'clear'
    force = rest.include?('-f')
    unless force
      print colorize("Clear all blocklist? [y/N] ", :yellow)
      ans = $stdin.gets.chomp
      return unless ans.downcase == 'y'
    end
    if cfg['blocked'].empty?
      puts colorize("Blocklist already empty.", :yellow)
      return
    end
    update_hosts(cfg['blocked'], 'remove')
    cfg['blocked'] = []
    cfg['active'] = false
    cfg['timer_end'] = nil
    save_config(cfg)
    puts colorize("Blocklist cleared.", :green)

  else
    puts colorize("Unknown command: #{cmd}", :red)
  end
end

main if __FILE__ == $0
